vernac/vernacentries.ml

```diff
@@ -2291,36 +2291,30 @@ let vernac_timeout ?timeout (f : 'a -> 'b) (x : 'a) : 'b =
     f x
 
 (* Fail *)
-exception HasNotFailed
-exception HasFailed of Pp.t
-
 let test_mode = ref false
 
-(* XXX STATE: this type hints that restoring the state should be the
-   caller's responsibility *)
-let with_fail ~st f =
+(* Restoring the state is the caller's responsibility *)
+let with_fail f : (Pp.t, unit) result =
   try
-    (* If the command actually works, ignore its effects on the state.
-       * Note that error has to be printed in the right state, hence
-       * within the purified function *)
-    try let _ = f () in raise HasNotFailed
-    with
-    | HasNotFailed as e -> raise e
-    | e when CErrors.noncritical e || e = Timeout ->
-      let e = CErrors.push e in
-      raise (HasFailed (CErrors.iprint (ExplainErr.process_vernac_interp_error e)))
-  with e when CErrors.noncritical e ->
-    (* Restore the previous state XXX Careful here with the cache! *)
-    Vernacstate.invalidate_cache ();
-    Vernacstate.unfreeze_interp_state st;
-    let (e, _) = CErrors.push e in
-    match e with
-    | HasNotFailed ->
-      user_err ~hdr:"Fail" (str "The command has not failed!")
-    | HasFailed msg ->
-      if not !Flags.quiet || !test_mode then Feedback.msg_info
-          (str "The command has indeed failed with message:" ++ fnl () ++ msg)
-    | _ -> assert false
+    let _ = f () in
+    Error ()
+  with
+  (* Fail Timeout is a common pattern so we need to support it. *)
+  | e when CErrors.noncritical e || e = Timeout ->
+    (* The error has to be printed in the failing state *)
+    Ok CErrors.(iprint ExplainErr.(process_vernac_interp_error (push e)))
```

Comment:
What happens if this raises an exception?

Reply:
An anomaly? Exception printing should not raise, right?

Reply:
For the record I did a test and I got:
```
File "./k.v", line 1, characters 0-23:
Error: Anomaly "Uncaught exception Vernacentries.Koo."
Please report at http://coq.inria.fr/bugs/.
```
where `Koo` is a hack so the fail printer raises.

Reply:
The exact code flow if you are curious is actually than once the printing has raised an exception, the cache will be invalidtated and the document manager will try to process the printing exception.

Usually the document manager won't be able to process this exception [for example a new found] so hence the anomaly will happen, but it could be well the case that I'd like to process the printing exn and turn into an user message.